middleware/ratelimiter.go

```diff
@@ -0,0 +1,511 @@
+		abstractLimiter
+		prefix string
+	}
+
+	// Result of limiter.Get
+	Result struct {
+		Total     int           // It Equals Options.Max, or policy max
+		Remaining int           // It will always >= -1
+		Duration  time.Duration // It Equals Options.Duration, or policy duration
+		Reset     time.Time     // The limit record reset time
+	}
+
+	abstractLimiter interface {
+		getLimit(key string, policy ...int) ([]interface{}, error)
+		removeLimit(key string) error
+	}
+
+	//RedisClient interface
+	RedisClient interface {
+		DeleteKey(string) error
+		EvalulateSha(string, []string, ...interface{}) (interface{}, error)
+		LuaScriptLoad(string) (string, error)
+	}
+
+)
+
+var (
+	// DefaultRateLimiterConfig is the default rate limit middleware config.
+	DefaultRateLimiterConfig = RateLimiterConfig{
+		Skipper:      DefaultSkipper,
+		LimitConfig:LimiterConfig{
+			Max:100,
+			Duration: time.Minute * 1,
+			Strategy:"ip",
+			Key:"",
+		},
+		Prefix:"LIMIT",
+		Client:nil,
+		SkipRateLimiterInternalError:false,
+	}
+	limiterImp *limiter
+)
+
+// RateLimiter returns a rate limit middleware.
+func RateLimiter() echo.MiddlewareFunc {
+	return RateLimiterWithConfig(DefaultRateLimiterConfig)
+}
+
+// RateLimiterWithConfig returns a RateLimiter middleware with config.
+// See: `RateLimiter()`.
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	// Defaults
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if len(config.LimitConfig.Strategy) == 0 {
+		panic("echo: rate limiter middleware requires Strategy (use ip or header(key)")
+	}
+	if config.Prefix == "" {
+		config.Prefix = "LIMIT:"
+	}
+	if config.LimitConfig.Max <= 0 {
+		config.LimitConfig.Max = 100
+	}
+	if config.LimitConfig.Duration <= 0 {
+		config.LimitConfig.Duration = time.Minute *1
+	}
+
+	//If config.Client omit, the limiter is a memory limiter
+	if config.Client == nil {
+		limiterImp = newMemoryLimiter(&config)
+	}else{
+		//setup redis client
+		limiterImp = newRedisLimiter(&config)
+	}
+
+	var tokenExtractor TokenExtractor
+
+	switch strings.ToLower(config.LimitConfig.Strategy) {
+	case "ip":
+		tokenExtractor = IPTokenExtractor()
+	case "header":
+		tokenExtractor = HeaderTokenExtractor(config.LimitConfig.Key)
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			response := c.Response()
+
+			policy := []int{}
+			result, err := limiterImp.Get(tokenExtractor(c),policy...)
+
+			if err != nil {
+
+				if config.SkipRateLimiterInternalError{
+					return next(c)
+				}
+				return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+			}
+
+			response.Header().Set("X-Ratelimit-Limit", strconv.FormatInt(int64(result.Total), 10))
+			response.Header().Set("X-Ratelimit-Remaining", strconv.FormatInt(int64(result.Remaining), 10))
+			response.Header().Set("X-Ratelimit-Reset", strconv.FormatInt(result.Reset.Unix(), 10))
```

Comment:
If you decide to support multiple instances, these header names could be modified like:

`fmt.Sprintf("X-Ratelimit-Remaining-%s", config.Prefix)`

Reply:
Prefix is used for the cache keys. We may use hostName instead of prefix.
something like;

```
hostName,_ :=  os.HostName()
fmt.Sprintf("X-Ratelimit-Remaining-%s", hostName)
```
or 
`fmt.Sprintf("X-Ratelimit-Host-%s", hostName)`
I think new response header more proper.

Reply:
It would be great if this patch implements the ratelimit I-D here https://tools.ietf.org/html/draft-polli-ratelimit-headers-01.

You just need to return a `delta-seconds` instead of a `unix-timestamp`. See comments for further infos.